substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/pecoff/cv/CVTypeSectionBuilder.java

```diff
@@ -0,0 +1,651 @@
+     *            at this time. This is used to avoid infinite recursion when processing
+     *            self-referencing types.
+     * @return The index for the typeentry for a pointer to the type.
+     */
+    int getIndexForPointerOrPrimitive(TypeEntry entry, boolean onlyForwardReference) {
+        if (entry.isPrimitive()) {
+            CVTypeRecord record = typeSection.getType(entry.getTypeName());
+            assert record != null;
+            return record.getSequenceNumber();
+        }
+        CVTypeRecord ptrRecord = typeSection.getPointerRecordForType(debugContext, entry.getTypeName());
+        if (ptrRecord == null) {
+            CVTypeRecord record = typeSection.getType(entry.getTypeName());
+            if (record == null) {
+                /* we've never heard of this class (but it may be in process) */
+                if (onlyForwardReference) {
+                    record = addTypeRecord(new CVTypeRecord.CVClassRecord((short) ATTR_FORWARD_REF, entry.getTypeName(), null));
+                    typeInfoMap.put(entry.getTypeName(), new TypeInfo(record, entry));
+                    log("getIndexForPointerOrPrimitive: type %s (%s): added %s", entry.getTypeName(), entry.typeKind().name(), record);
+                } else {
+                    record = buildType(entry);
+                }
+            } else if (record.getSequenceNumber() <= MAX_PRIMITIVE) {
+                /* Primitive types are referenced directly */
+                return record.getSequenceNumber();
+            }
+            /* We now have a class record but must create a pointer record. */
+            ptrRecord = addTypeRecord(new CVTypeRecord.CVTypePointerRecord(record.getSequenceNumber(), CVTypeRecord.CVTypePointerRecord.NORMAL_64));
+        }
+        return ptrRecord.getSequenceNumber();
+    }
+
+    /**
+     * If the type of the pointee is a primitive type, return it directly. Otherwise, create (if
+     * needed) and return a record representing a pointer to class.
+     *
+     * private int getIndexForPointer(String typeName) { CVTypeRecord ptrRecord =
+     * typeSection.getPointerRecordForType(typeName); if (ptrRecord == null) { CVTypeRecord
+     * clsRecord = typeSection.getType(typeName); if (clsRecord == null) { /* we've never heard of
+     * this class (but it may be in process) * clsRecord = addTypeRecord(new
+     * CVTypeRecord.CVClassRecord((short) ATTR_FORWARD_REF, typeName, null));
+     * typeInfoMap.put(typeName, new TypeInfo(clsRecord, null)); } else if
+     * (clsRecord.getSequenceNumber() <= MAX_PRIMITIVE) { return clsRecord.getSequenceNumber(); } /*
+     * We now have a class record but must create a pointer record. * ptrRecord = addTypeRecord(new
+     * CVTypeRecord.CVTypePointerRecord(clsRecord.getSequenceNumber(),
+     * CVTypeRecord.CVTypePointerRecord.NORMAL_64)); } return ptrRecord.getSequenceNumber(); }
+     */
```

Comment:
This comment really needs removing. If you really want to keep the code you can mark it as unused.

Reply:
Removed - it was mangled and I didn't catch it.  It was only required if there was such a thing as a pointer to a primitive non-array type in Java - which there isn't.